lighthouse-core/report/v2/renderer/report-features.js

```diff
@@ -0,0 +1,273 @@
+   * @param {ReportJSON=} report
+   */
+  attach(report) {
+    this.json = report;
+    this._addEventListeners();
+  }
+
+  /**
+   * Handler copy events.
+   */
+  onCopy(e) {
+    // Only handle copy button presses (e.g. ignore the user copying page text).
+    if (this._copyAttempt) {
+      // We want to write our own data to the clipboard, not the user's text selection.
+      e.preventDefault();
+      e.clipboardData.setData('text/plain', JSON.stringify(this.json, null, 2));
+      this.logger.log('Report JSON copied to clipboard');
+    }
+
+    this._copyAttempt = false;
+  }
+
+  /**
+   * Copies the report JSON to the clipboard (if supported by the browser).
+   */
+  onCopyButtonClick() {
+    if (window.ga) {
+      window.ga('send', 'event', 'report', 'copy');
+    }
+
+    try {
+      if (this._document.queryCommandSupported('copy')) {
+        this._copyAttempt = true;
+
+        // Note: In Safari 10.0.1, execCommand('copy') returns true if there's
+        // a valid text selection on the page. See http://caniuse.com/#feat=clipboard.
+        const successful = this._document.execCommand('copy');
+        if (!successful) {
+          this._copyAttempt = false; // Prevent event handler from seeing this as a copy attempt.
+          this.logger.warn('Your browser does not support copy to clipboard.');
+        }
+      }
+    } catch (err) {
+      this._copyAttempt = false;
+      this.logger.log(err.message);
+    }
+  }
+
+  closeExportDropdown() {
+    this.exportButton.classList.remove('active');
+  }
+
+  /**
+   * Click handler for export button.
+   */
+  onExportButtonClick(e) {
+    e.preventDefault();
+    e.target.classList.toggle('active');
+    this._document.addEventListener('keydown', this.onKeyDown);
+  }
+
+  /**
+   * Handler for "export as" button.
+   */
+  onExport(e) {
+    e.preventDefault();
+
+    if (!e.target.dataset.action) {
+      return;
+    }
+
+    switch (e.target.dataset.action) {
+      case 'copy':
+        this.onCopyButtonClick();
+        break;
+      case 'open-viewer':
+        this.sendJSONReport();
+        break;
+      case 'print':
+        this.expandDetailsWhenPrinting();
+        window.print();
+        break;
+      case 'save-json': {
+        const jsonStr = JSON.stringify(this.json, null, 2);
+        this._saveFile(new Blob([jsonStr], {type: 'application/json'}));
+        break;
+      }
+      case 'save-html': {
+        let htmlStr = '';
+
+        // Since Viewer generates its page HTML dynamically from report JSON,
+        // run the ReportGenerator. For everything else, the page's HTML is
+        // already the final product.
+        // if (e.target.dataset.context !== 'viewer') {
+        //   htmlStr = this._document.documentElement.outerHTML;
+        // } else {
+        //   const reportGenerator = new ReportGeneratorV2();
+        //   htmlStr = reportGenerator.generateReportHTML(this.json);
+        // }
+        // TODO: fix viewer.
```

Comment:
when we do, instead of generating it on the fly I'd be interested to publish the bundled JS with our releases instead that the viewer can just reference. This way the viewer can actually start supporting all versions going forward too